### PR TITLE
importer: rewrite ts for download SST

### PR DESCRIPTION
**Pull request by Connor1996**
###  What have you changed?
Part of #6222 rewrite ts for download SSTs when doing the incremental restore

Should merge #6219 first

###  What is the type of the changes?
- New feature (a change which adds functionality)

###  How is the PR tested?
- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No

###  Does this PR affect `tidb-ansible`?
No

### CODE REVIEW

File: components/sst_importer/src/sst_importer.rs [hunk 1]
```diff
@@ -231,26 +233,55 @@ impl SSTImporter {
             Bound::Excluded(_) => unreachable!(),
         };
         while iter.valid() {
-            let old_key = keys::origin_key(iter.key());
+            let mut k = Key::from_encoded_slice(keys::origin_key(iter.key()));
```
Comment (overvenus) [inline components/sst_importer/src/sst_importer.rs]: Could you create encoded_slce lazily? It allocates memory.

File: components/sst_importer/src/sst_importer.rs [hunk 2]
```diff
@@ -231,26 +233,55 @@ impl SSTImporter {
...
+            if rewrite_rule.new_timestamp != 0 && meta.get_cf_name() == "write" {
+                let mut write = WriteRef::parse(iter.value()).map_err(|e| {
+                    Error::BadFormat(format!(
+                        "write {}: {}",
+                        hex::encode_upper(keys::origin_key(iter.key()).to_vec()),
+                        e
+                    ))
+                })?;
+                write.start_ts = TimeStamp::new(rewrite_rule.new_timestamp);
+                sst_writer.put(&key, &write.to_bytes())?
```
Comment (overvenus) [inline components/sst_importer/src/sst_importer.rs]: What about commit ts in the key? Is it possible that `commit_ts <= start_ts`?

Comment (Connor1996) [reply components/sst_importer/src/sst_importer.rs]: the `commit_ts` is also rewritten to `new_timestamp`

File: components/sst_importer/src/sst_importer.rs [hunk 3]
```diff
@@ -231,26 +235,58 @@ impl SSTImporter {
             Bound::Excluded(_) => unreachable!(),
         };
         while iter.valid() {
-            let old_key = keys::origin_key(iter.key());
-            if is_after_end_bound(old_key, &range_end) {
+            let old_key = {
+                if rewrite_rule.new_timestamp != 0 {
+                    Cow::from(
```
Comment (5kbpers) [inline components/sst_importer/src/sst_importer.rs]: Is `Cow` necessary?  Seems the `old_key` would never be write.

Comment (Connor1996) [reply components/sst_importer/src/sst_importer.rs]: it is used to unify `Vec<u8>` and `&[u8]`, or it needs extra memory allocation for `&[u8]`

File: components/sst_importer/src/sst_importer.rs (hunk 2 above)
Comment (kennytm) [reply components/sst_importer/src/sst_importer.rs]: If I were to write this code I'd make it like this:

```rust
        let is_write_cf = meta.get_cf_name() == "write";
        let new_ts = TimeStamp::new(rewrite_rule.new_timestamp);

        while iter.valid() {
            // Don't rewrite anything while we're still checking if old_key is valid.
            let old_key = keys::origin_key(iter.key());
            if is_after_end_bound(old_key, &range_end) {
                break;
            }
            if !old_key.starts_with(old_prefix) {
                return Err(Error::WrongKeyPrefix(
                    "Key in SST",
                    old_key.to_vec(),
                    old_prefix.to_vec(),
                ));
            }

            key.truncate(new_prefix_data_key_len);
            key.extend_from_slice(&old_key[old_prefix.len()..]);
            let mut value = Cow::Borrowed(iter.value());

            // rewrite the _new_ key.
            if !new_ts.is_zero() {
                // consider introducing `Key::rewrite_ts_for(&mut [u8], TimeStamp)` 
                // so we don't need to adjust the length but directly rewrite the last 8 bytes
                key = Key::from_encoded(key)
                    .truncate_ts()
                    .map(|k| k.append_ts(new_ts))
                    .map_err(|e| {
                        Error::BadFormat(format!("key {}: {}", hex::encode_upper(old_key)), e)
                    })?
                    .into_encoded();
                if is_write_cf {
                    let mut write = WriteRef::parse(value).map_err(|e| {
                        Error::BadFormat(format!("value {}: {}", hex::encode_upper(value), e))
                    })?;
                    write.start_ts = new_ts;
                    value = Cow::Owned(write.to_bytes());
                }
            }

            sst_writer.put(&key, &value)?;
            iter.next();
            if first_key.is_none() {
                first_key = Some(keys::origin_key(&key).to_vec());
            }
        }
```

File: components/sst_importer/src/sst_importer.rs (hunk 3 above)
Comment (kennytm) [reply components/sst_importer/src/sst_importer.rs]: If I were to write this code I'd make it like this:

```rust
        let is_write_cf = meta.get_cf_name() == "write";
        let new_ts = TimeStamp::new(rewrite_rule.new_timestamp);

        while iter.valid() {
            // Don't rewrite anything while we're still checking if old_key is valid.
            let old_key = keys::origin_key(iter.key());
            if is_after_end_bound(old_key, &range_end) {
                break;
            }
            if !old_key.starts_with(old_prefix) {
                return Err(Error::WrongKeyPrefix(
                    "Key in SST",
                    old_key.to_vec(),
                    old_prefix.to_vec(),
                ));
            }

            key.truncate(new_prefix_data_key_len);
            key.extend_from_slice(&old_key[old_prefix.len()..]);
            let mut value = Cow::Borrowed(iter.value());

            // rewrite the _new_ key.
            if !new_ts.is_zero() {
                // consider introducing `Key::rewrite_ts_for(&mut [u8], TimeStamp)` 
                // so we don't need to adjust the length but directly rewrite the last 8 bytes
                key = Key::from_encoded(key)
                    .truncate_ts()
                    .map(|k| k.append_ts(new_ts))
                    .map_err(|e| {
                        Error::BadFormat(format!("key {}: {}", hex::encode_upper(old_key)), e)
                    })?
                    .into_encoded();
                if is_write_cf {
                    let mut write = WriteRef::parse(value).map_err(|e| {
                        Error::BadFormat(format!("value {}: {}", hex::encode_upper(value), e))
                    })?;
                    write.start_ts = new_ts;
                    value = Cow::Owned(write.to_bytes());
                }
            }

            sst_writer.put(&key, &value)?;
            iter.next();
            if first_key.is_none() {
                first_key = Some(keys::origin_key(&key).to_vec());
            }
        }
```

File: components/sst_importer/src/sst_importer.rs [hunk 4]
```diff
@@ -232,25 +236,52 @@ impl SSTImporter {
...
+                    .map_err(|e| {
+                        Error::BadFormat(format!(
+                            "key {}: {}",
+                            hex::encode_upper(keys::origin_key(iter.key()).to_vec()),
+                            e
+                        ))
+                    })?
+                    .append_ts(TimeStamp::new(rewrite_rule.new_timestamp))
+                    .into_encoded();
+                if meta.get_cf_name() == "write" {
```
Comment (overvenus) [inline components/sst_importer/src/sst_importer.rs]: How about comparing it with `engine_trait::CF_WRITE`?

**Comment by Connor1996** [comment]
PTAL again @overvenus @5kbpers

**Pull request merged**